hilti/toolchain/src/compiler/optimizer/optimizer.cc

```diff
@@ -0,0 +1,434 @@
+                                    util::fmt("  * deleting function state: %s (via module %s)", function->id(),
+                                              module->id()));
+                }
+            }
+        }
+
+        if ( ! modified )
+            break;
+
+        HILTI_DEBUG(logging::debug::Optimizer, "  -> modified");
+
+        if ( ++round >= ASTContext::MaxASTIterationRounds )
+            logger().internalError("Optimizer::_updateState() didn't terminate, AST keeps changing");
+    }
+
+    _modified_functions.clear();
+    _modified_modules.clear();
+}
+
+#ifndef NDEBUG
+void ASTState::checkAST(std::string_view pass_name) {
+    // In debug builds, we check the AST after each pass to enforce that it's
+    // been left in good shape.
+    util::timing::Collector _("hilti/compiler/optimizer/check-state");
+
+    context()->checkAST();
+
+    validator::detail::validatePost(builder(), context()->root());
+    if ( ! context()->collectErrors() )
+        logger().internalError("Optimizer::_checkState: AST is not valid after optimizer pass");
+
+    if ( ! type_unifier::check(builder(), context()->root()) )
+        logger().internalError(
+            util::fmt("Optimizer::_checkState: AST types are not fully unified after optimizer pass %s", pass_name));
+
+    if ( scope_builder::build(builder(), context()->root()) )
+        logger().internalError(
+            util::fmt("Optimizer::_checkState: AST scopes are not fully built after optimizer pass %s", pass_name));
+
+    // We check folding here without the any additional styles otherwise used
+    // inside the optimizer, because that's what the normal resolver does. If
+    // we checked for them here, we could trigger in case the original AST,
+    // which only went through that standard resolving, gets here unmodified.
+    if ( constant_folder::fold(builder(), context()->root()) )
+        logger().internalError(
+            util::fmt("Optimizer::_checkState: AST is not fully constant folded after optimizer pass %s", pass_name));
+
+    if ( resolver::coerce(builder(), context()->root()) )
+        logger().internalError(
+            util::fmt("Optimizer::_checkState: AST is not fully coerced after optimizer pass %s", pass_name));
+
+    if ( resolver::resolve(builder(), context()->root()) )
+        logger().internalError(
+            util::fmt("Optimizer::_checkState: AST is not fully resolved after optimizer pass %s", pass_name));
+
+    // Go through cached CFGs and verify that they are still up to date.
+    for ( const auto& [block, cfg] : _cfgs ) {
+        assert(block);
+
+        if ( ! cfg )
+            continue; // block no longer part of AST
+
+        auto actual = cfg->dot(false);
+        auto expected = CFG(block).dot(false);
+
+        if ( actual != expected ) {
+            std::cerr << "==== ACTUAL   ===\n\n" << actual << "\n\n";
+            std::cerr << "==== EXPECTED ===\n\n" << expected << "\n\n";
```

Comment:
nit: LHS has 4 equals and RHS has 3! (I only spotted because my editor has a special glyph for 3 equals, not 4)


```suggestion
            std::cerr << "=== ACTUAL   ===\n\n" << actual << "\n\n";
            std::cerr << "=== EXPECTED ===\n\n" << expected << "\n\n";
```